### ISSUE TITLE
Chain of Trust for SEV-SNP

### ISSUE THREAD

**Issue by fitzthum** [confidential-containers/community#54]
For SEV-SNP the measurement included in the attestation report only covers flash0 (at least for QEMU). For confidential containers, we need to measure the firmware, the initrd, the kernel, and the kernel command line before releasing secrets to a guest. How can we use the attestation report to measure all of these components? 

One option is "measured direct boot," which we currently use for CCv0 with SEV. Measured direct boot relies on the[ AmdSev OVMF package](https://github.com/tianocore/edk2/blob/master/OvmfPkg/AmdSev/AmdSevX64.dsc), which sets aside a special piece of the firmware binary for storing the hashes of the initrd, kernel, and command line. When QEMU boots a VM, it hashes each of these components and injects the hashes into the firmware binary. In a direct boot OVMF receives each of these components from the HV via the fw_config interface. When OVMF receives one of these components from the HV, OVMF measures the component and compares the measurement to the measurement injected into the firmware binary. if the measurement does not match, the boot fails. The HV is untrusted, but fw binary is measured, so the hashes that the HV injects can be verified by the guest owner. Support for measured direct boot for SEV(-ES) is already upstream in OVMF and QEMU. Some minor changes are needed to support SEV-SNP. These changes have been documented [here](https://edk2.groups.io/g/devel/message/88137)

There are a number of possible variations to measured direct boot, such as using host data to provide the expected measurements. We can discuss the tradeoffs here or on the OVMF list.

There are other ways to securely boot with SEV-SNP. There are some approaches that provide an encrypted root disk. In the past we have opted against using encrypted disk (long discussion [here](https://github.com/confidential-containers/confidential-containers/issues/116), but we could reconsider. I don't know as much about these approaches so I will leave it to someone else to elaborate. 

At some point there might also be support for secure vTPMs in SEV-SNP. I think a vTPM might be overkill for confidential containers, but we could potentially implement a TPM KBC to make use of one.

Finally, it's worth noting that for peer pods (where the shim calls a CSP's API to launch the pod vm) things could get a bit messy if different CSPs decide to answer this question in different ways.

**Comment by dubek** [comment]
> There are a number of possible variations to measured direct boot, such as using host data to provide the expected measurements.

1. Note that host data is only 256 bits long.
2. I think the only way the guest can read the host data is by asking the PSP for the attestation report; for this we need to implement the SNP guest message request protocol inside OVMF. That means: reading VMPCK keys from the SNP secrets page, encrypting/decryption messages to/from the PSP with AES-GCM-256, parsing the messages -- everything in the guest kernel's drivers/virt/coco/sev-guest driver.

**Comment by jepio** [comment]
> When QEMU boots a VM, it hashes each of these components and injects the hashes into the firmware binary. In a direct boot OVMF receives each of these components from the HV via the fw_config interface. 

The individual hashes also need to make their way to the guest OS somehow, to be provided as part of "evidence" by KBC to KBS. This applies regardless of how exactly the hashes are included in the launch measurement (host_data vs. "measured direct boot"). When performing direct kernel boot, the vmlinuz/initrd files are not part of /boot and I don't see that they are accessible to the guest OS, so we'll have to find a way to pass that information in.

**Comment by bodzhang** [comment]
Interesting [blog](https://decentriq.notion.site/decentriq/Swiss-cheese-to-cheddar-securing-AMD-SEV-SNP-early-boot-9927b637d4914745a907f04bc651110b) from decentriq about their effort to use a bare minimum firmware to "direct boot" the Linux kernel (small) and initrd in SEV-SNP VM, with the HW measuring everything.

**Comment by fitzthum** [comment]
At first glance this seems fairly similar to measured direct boot. They don't give much info about how the firmware validates the data that it gets via `fw_cfg`, but it's definitely still a two-stage measurement where the firmware measures the kernel. It's kind of nifty that they use a combined initrd, kernel, cmdline image although that isn't necessarily very useful for Kata given that those components are supplied separately. 

I think using something like qboot rather than OVMF is potentially a good idea. It requires extensions to the firmware and some changes to QEMU. Rather than qboot I think we should probably be looking at adding SEV-SNP support to [td-shim ](https://github.com/confidential-containers/td-shim).

**Comment by bodzhang** [comment]
I believe in decentriq's approach, the kernel is part of the "firmware" and directly measured by SEV-SNP - _"kernel, kernel cmdline, initrd: This is the operating system itself to be loaded by the firmware. The way we eliminated this input was by burning all of these into the firmware blob itself (so we don’t even need to load it and measure it). The only tricky part here was ensuring that the kernel actually fit the maximum size of the firmware. The kernel we embedded was minimal from the get go, for security purposes we based it on the tinyconfig configuration. But we also had to limit initrd, which is a bit trickier as the workload compute can be a lot larger. We solved this by introducing a second boot stage. The initial stage 1 initrd contains only a statically compiled busybox, and once it starts it loads and measures a stage 2 initrd through virtio VSOCK, which is then mounted and chrooted. This way the final combined firmware fit into less than 4MB, which incidentally also made VM boots very fast."_

Adding SEV-SNP support in [td-shim ](https://github.com/confidential-containers/td-shim) would be great.

**Comment by fitzthum** [comment]
> I believe in decentriq's approach, the kernel is part of the "firmware" and directly measured by SEV-SNP

Yeah that's right. I guess they still do a two-stage measurement, but they don't have the firmware measure the kernel. Rather they have a minimal kernel that can then measure a larger second stage initrd. I think some kind of two-stage approach might be inevitable given some of the size limitations. The more stuff you measure via the PSP the longer the boot will take. I'm not sure that splitting the phases into firmware+initrd1 and initrd2 is necessarily better than doing it between firmware and initrd, especially given how Kata works today.

**Comment by dubek** [comment]
One more thing to consider: as part of decentriq's quest to reduce untrusted communication with the hypervisor:

> To eliminate this input we decided to fix the number of vCPUs during build time and burn this number into the metadata.

This seems too rigid too me: you'll need a different pre-built image (= firmware+kernel+initrd1) if you want to start a VM with different number of CPUs. There might be similar issues with amount of RAM or adding/removing other devices.

I think we should eventually find a way for these to be configurable (so we don't have to pre-build a binary for each configuration) **but** easily measured (it's not so easy :-) ).

**Comment by ariel-adam** [comment]
@fitzthum is this issue still relevant or can be closed? 
If it's still relevant to what release do you think we should map it to (mid-November, end-December, mid-February etc...)?

**Comment by zvonkok** [comment]
Any ideas on how we can extend the chain of trust of additional components that are needed to run a confidential container? 
The first step laid out here is for a pure CPU-based workload. In the case of GPUs, we need to deploy a custom kernel and guest-fs. Additionally, we would have a custom attestation report of the GPU stack (firmware, VBIOS, GSP-API, ... )  that needs to be attested with the rest of the system as the TCB.
